source/common/network/connection_impl.cc

```diff
@@ -403,6 +430,28 @@ void ConnectionImpl::onFileEvent(uint32_t events) {
     return;
   }
 
+#ifdef __APPLE__
+  // Because OSX doesn't support early close detection via events, we leave the read event enabled
+  // even when reads are disabled. Detect it here and convert it to a Closed event if we can detect
+  // the connection is closed. See https://github.com/envoyproxy/envoy/issues/4294.
+  if (detect_early_close_ && !read_enabled_ && (events & Event::FileReadyType::Read) != 0) {
+    if (detectEarlyClose()) {
+      // No longer connected. Convert to a closed event.
+      ENVOY_CONN_LOG(trace, "early close detection triggered", *this);
+      events |= Event::FileReadyType::Closed;
```

Comment:
shouldn't you remove the read regardless?

Reply:
That's done further down, outside the if/else.